File: api/src/main/java/net/blueshell/api/listener/email/RecoveryEmailEventListener.java
package net.blueshell.api.listener.email;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.blueshell.api.common.event.job.RecoveryEmailEvent;
import net.blueshell.api.job.email.RecoveryEmailJob;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class RecoveryEmailEventListener {

    private final RecoveryEmailJob recoveryEmailJob;

    @EventListener
    public void onReset(RecoveryEmailEvent evt) {
        var userId = evt.userId();
        if (evt.userId() == null) return;

        recoveryEmailJob.send(userId, evt.token(), evt.resetType());
    }
}

File: api/src/main/kotlin/net/blueshell/api/listener/email/RecoveryEmailEventListener.kt
package net.blueshell.api.listener.email

import net.blueshell.api.common.event.job.RecoveryEmailEvent
import net.blueshell.api.job.email.RecoveryEmailJob
import org.springframework.stereotype.Component
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener

@Component
class RecoveryEmailEventListener(
    private val recoveryEmailJob: RecoveryEmailJob
) {
    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    fun onReset(evt: RecoveryEmailEvent) {
        val userId = evt.userId
        if (evt.userId == null) return

        recoveryEmailJob.send(userId, evt.token, evt.resetType)
    }
}
